libsolidity/ast/Types.cpp

```diff
@@ -362,6 +369,74 @@ MemberList::MemberMap Type::boundFunctions(Type const& _type, ASTNode const& _sc
 	if (auto refType = dynamic_cast<ReferenceType const*>(&_type))
 		typeLocation = refType->location();
 
+	return usingForDirectives | ranges::views::filter([&](UsingForDirective const* _directive) -> bool {
+		// Convert both types to pointers for comparison to see if the `using for` directive applies.
+		// Note that at this point we don't yet know if the functions are actually usable with the type.
+		// `_type` may not be convertible to the function parameter type.
+		return
+			!_directive->typeName() ||
+			*TypeProvider::withLocationIfReference(typeLocation, &_type, true) ==
+			*TypeProvider::withLocationIfReference(
+				typeLocation,
+				_directive->typeName()->annotation().type,
+				true
+			);
+	}) | ranges::to<vector<UsingForDirective const*>>;
+}
+
+}
+
+Result<FunctionDefinition const*> Type::userDefinedOperator(Token _token, ASTNode const& _scope, bool _unaryOperation) const
+{
+	if (!typeDefinition())
+		return nullptr;
+
+	set<FunctionDefinition const*> matchingDefinitions;
+	for (UsingForDirective const* ufd: usingForDirectivesForType(*this, _scope))
+		for (auto const& [identifierPath, operator_]: ufd->functionsAndOperators())
+		{
+			if (operator_ != _token)
+				continue;
+			FunctionDefinition const& function = dynamic_cast<FunctionDefinition const&>(
+				*identifierPath->annotation().referencedDeclaration
+			);
+			FunctionType const* functionType = dynamic_cast<FunctionType const*>(
+				function.libraryFunction() ? function.typeViaContractName() : function.type()
+			);
+			solAssert(functionType && !functionType->parameterTypes().empty());
+
+			Type const* normalizedType = this;
+			if (auto const* referenceType = dynamic_cast<ReferenceType const*>(normalizedType))
+				normalizedType = TypeProvider::withLocationIfReference(referenceType->location(), normalizedType);
+
+			Type const* normalizedParameterType = functionType->parameterTypes().front();
+			if (auto const* referenceType = dynamic_cast<ReferenceType const*>(normalizedParameterType))
+				normalizedParameterType = TypeProvider::withLocationIfReference(referenceType->location(), normalizedParameterType);
+
+			if (
+				*normalizedType == *normalizedParameterType &&
+				(
+					(_unaryOperation && function.parameterList().parameters().size() == 1) ||
+					(!_unaryOperation && function.parameterList().parameters().size() == 2)
+				)
+			)
+				matchingDefinitions.insert(&function);
```

Comment:
Since we expect that at this point the directives have been validated by the type checker, it would be simpler to expect at least one argument in all cases:

```suggestion
			if (*normalizedType == *normalizedParameterType && function.parameterList().parameters().size() >= 1)
				matchingDefinitions.insert(&function);
```
We're not even checking the other arguments and it's not our job here. I think this function should simply return whatever the type checker let through.

Reply:
This code is not only to confirm that a function with two input parameters is bound to a binary operator and a function with one parameter to a unary operator. Notice that `_unaryOperation` is the function argument and it is used to pick up the right function definition. Without that, the function would always return all the functions bound to the operator regardless a function for unary or binary operator is requested. That would also require an extra filtering on the caller side.

Reply:
ok, you're right. This makes sense.